core/trino-main/src/main/java/io/trino/execution/scheduler/FaultTolerantStageScheduler.java

```diff
@@ -0,0 +1,530 @@
+        }
+        return result.build();
+    }
+
+    private void updateTaskStatus(TaskStatus taskStatus, Optional<ExchangeSinkInstanceHandle> exchangeSinkInstanceHandle)
+    {
+        TaskState state = taskStatus.getState();
+        if (!state.isDone()) {
+            return;
+        }
+
+        try {
+            RuntimeException failure = null;
+            SettableFuture<Void> future;
+            synchronized (this) {
+                TaskId taskId = taskStatus.getTaskId();
+
+                runningTasks.remove(taskId);
+                future = taskFinishedFuture;
+                if (!runningTasks.isEmpty()) {
+                    taskFinishedFuture = SettableFuture.create();
+                }
+                else {
+                    taskFinishedFuture = null;
+                }
+
+                InternalNode node = requireNonNull(runningNodes.remove(taskId), () -> "node not found for task id: " + taskId);
+                nodeAllocator.release(node);
+
+                int partitionId = taskId.getPartitionId();
+
+                if (!finishedPartitions.contains(partitionId) && !closed) {
+                    switch (state) {
+                        case FINISHED:
+                            finishedPartitions.add(partitionId);
+                            if (sinkExchange.isPresent()) {
+                                checkArgument(exchangeSinkInstanceHandle.isPresent(), "exchangeSinkInstanceHandle is expected to be present");
+                                sinkExchange.get().sinkFinished(exchangeSinkInstanceHandle.get());
+                            }
+                            partitionToRemoteTaskMap.get(partitionId).forEach(RemoteTask::abort);
+                            break;
+                        case CANCELED:
+                            log.info("Task cancelled: %s", taskId);
+                            break;
+                        case ABORTED:
+                            log.info("Task aborted: %s", taskId);
+                            break;
+                        case FAILED:
+                            ExecutionFailureInfo failureInfo = taskStatus.getFailures().stream()
+                                    .findFirst()
+                                    .map(f -> rewriteTransportFailure(failureDetector, f))
+                                    .orElse(toFailure(new TrinoException(GENERIC_INTERNAL_ERROR, "A task failed for an unknown reason")));
+                            log.warn(failureInfo.toException(), "Task failed: %s", taskId);
+                            ErrorCode errorCode = failureInfo.getErrorCode();
+                            if (remainingRetryAttempts > 0 && (errorCode == null || errorCode.getType() != USER_ERROR)) {
+                                remainingRetryAttempts--;
+                                // schedule failed tasks first
```

Comment:
If we want to schedule failed task first, should we declare as a Deque so we insert `partitionId` to the head?

Reply:
It used to be that I thought it might be better to retry first. But then I thought it might be better quite the opposite (e.g.: if something failed due to infrastructure problems it feels it is better to give it some time to recover while processing other tasks). But I don't have a strong opinion here. We might want to change it at we move forward. For now removed the comment to avoid confusion.